File: tools/ioflow/ioflow-syncwrite.c
/* 
 * ioflow-syncwrite.c        C source code for sync write tracer. 
 *
 *
 * Copyright (c) Google LLC
 * Licensed under the Apache License, Version 2.0 (the "License")
 *
 * 27-Aug-2020   Haoning Chen   Created this.
 */

#include <uapi/linux/ptrace.h>
#include <linux/blkdev.h>
#include <linux/fs.h>

#define IO_FLOW_SYNCWRITE 1
// Use python string replacement to import common header code
[IMPORT_COMM]

BPF_HASH(syscall_map, u64, struct val_t);    // Map of syscall data, with pid as key
BPF_HASH(request_map, struct request *, struct rqval_t);    // Map of request info for asynchronous request handling

BPF_HISTOGRAM(hist_vfs);
BPF_HISTOGRAM(hist_ext4);
BPF_HISTOGRAM(hist_writepg);
BPF_HISTOGRAM(hist_ext4sync);
BPF_HISTOGRAM(hist_blk);
BPF_HISTOGRAM(hist_split);
BPF_HISTOGRAM(hist_merge);
BPF_HISTOGRAM(hist_rq_queue);
BPF_HISTOGRAM(hist_rq_service);

BPF_PERF_OUTPUT(syscall_events);
BPF_PERF_OUTPUT(rq_events);


int vfs_write_entry(struct pt_regs *ctx, struct file *file) {
    // Filter asyc and direct IO
    if (!(file->f_op->write_iter) || file->f_flags & O_DIRECT)
        return 0;
    
    u64 pid =  bpf_get_current_pid_tgid();
    struct val_t val = {0};
    comm_vfs_entry(&val, file);
    syscall_map.update(&pid, &val);
    return 0;
}

int ext4_entry(struct pt_regs *ctx) {
    u64 pid =  bpf_get_current_pid_tgid();
    struct val_t *valp = syscall_map.lookup(&pid);
    if (valp && valp->ts_ext4 == 0)
        valp->ts_ext4 = bpf_ktime_get_ns();
    return 0;
}

int write_page_entry(struct pt_regs *ctx) {
    u64 pid =  bpf_get_current_pid_tgid();
    struct val_t *valp = syscall_map.lookup(&pid);
    if (valp && valp->ts_writepg == 0)
        valp->ts_writepg = bpf_ktime_get_ns();
    return 0;
}

int ext4_sync_entry(struct pt_regs *ctx) {
    u64 pid =  bpf_get_current_pid_tgid();
    struct val_t *valp = syscall_map.lookup(&pid);
    if (valp && valp->ts_ext4sync == 0)
        valp->ts_ext4sync = bpf_ktime_get_ns();
    return 0;
}

/* BLock layer */
int block_entry(struct pt_regs *ctx) {
    u64 ts = bpf_ktime_get_ns();
    u64 pid = bpf_get_current_pid_tgid();
    struct val_t *valp = syscall_map.lookup(&pid);
    if (valp) {
        comm_block_entry(valp, ts);
    }
    return 0;
}

int block_return(struct pt_regs *ctx) {
    u64 ts = bpf_ktime_get_ns();
    u64 pid = bpf_get_current_pid_tgid();
    struct val_t *valp = syscall_map.lookup(&pid);
    if (valp) {
        comm_block_return(valp, ts);
    }
    return 0;
}

int split_entry(struct pt_regs *ctx) {
    u64 ts = bpf_ktime_get_ns();
    u64 pid = bpf_get_current_pid_tgid();
    struct val_t *valp = syscall_map.lookup(&pid);
    if (valp) {
        comm_split_entry(valp, ts);
    }
    return 0;
}

int split_return(struct pt_regs *ctx) {
    u64 ts = bpf_ktime_get_ns();
    u64 pid = bpf_get_current_pid_tgid();
    struct val_t *valp = syscall_map.lookup(&pid);
    if (valp) {
        comm_split_return(valp, ts);
    }
    return 0;
}

int merge_entry(struct pt_regs *ctx) {
    u64 ts = bpf_ktime_get_ns();
    u64 pid = bpf_get_current_pid_tgid();
    struct val_t *valp = syscall_map.lookup(&pid);
    if (valp) {
        comm_merge_entry(valp, ts);
    }
    return 0;
}

int merge_return(struct pt_regs *ctx) {
    u64 ts = bpf_ktime_get_ns();
    u64 pid = bpf_get_current_pid_tgid();
    struct val_t *valp = syscall_map.lookup(&pid);
    if (valp) {
        comm_merge_return(valp, ts);
    }
    return 0;
}

// Async request handling
int rq_create(struct pt_regs *ctx, struct request *rq) {
    // Still in the syscall process's context now. 
    u64 ts = bpf_ktime_get_ns();
    u64 pid = bpf_get_current_pid_tgid();
    struct val_t *valp = syscall_map.lookup(&pid);
    if (valp) {
        struct rqval_t rqval = {0};
        comm_rq_create(&rqval, valp, rq, pid >> 32, ts);
        request_map.insert(&rq, &rqval);
    }
    return 0;
}

// The request is issued to device driver.
int rq_issue(struct pt_regs *ctx, struct request *rq) {
    // Async to the syscall process now. 
    u64 ts = bpf_ktime_get_ns();
    struct rqval_t *rqvalp = request_map.lookup(&rq);
    comm_rq_issue(rqvalp, ts);
    return 0;
}

// The request is done.
int rq_done(struct pt_regs *ctx, struct request *rq) {
    u64 ts = bpf_ktime_get_ns();
    struct rqval_t *rqvalp = request_map.lookup(&rq);
    if (rqvalp) {
        struct rqdata_t rqdata = {0};
        comm_rq_done(&rqdata, rqvalp, ts);
        rq_events.perf_submit(ctx, &rqdata, sizeof(rqdata));

        hist_rq_queue.increment(bpf_log2l(rqdata.queue / 1000));
        hist_rq_service.increment(bpf_log2l(rqdata.service / 1000));
        
        request_map.delete(&rq);
    }
    return 0;
}

// The end of the write IO
int vfs_write_return(struct pt_regs *ctx) {
    struct file *file = (struct file *) PT_REGS_PARM1(ctx);
    ssize_t size = (ssize_t) PT_REGS_RC(ctx);
    u64 pid =  bpf_get_current_pid_tgid();

    struct val_t *valp = syscall_map.lookup(&pid);
    if (!valp) 
        return 0;
    if (!valp->ts_ext4 || !valp->ts_writepg) {
        // filter out non-buffered wirtes
        syscall_map.delete(&pid);
        return 0;
    }

    /* Note: the async (normal) writes aren't filtered. In those 
    *  cases, timestamp fields starting from ext4sync are all 0. 
    */

    // Populate output data struct.
    u64 ts_end = bpf_ktime_get_ns();
    struct data_t data = {0};
    comm_vfs_return(&data, valp, pid >> 32, ts_end, size);
    
    data.vfs = valp->ts_ext4 - valp->ts_vfs;
    data.ext4 = valp->ts_writepg - valp->ts_ext4;
    if (valp->ts_ext4sync > valp->ts_writepg)
        data.writepg = valp->ts_ext4sync - valp->ts_writepg;
    else
        data.writepg = ts_end - valp->ts_writepg;

    if (valp->ts_ext4sync != 0 && valp->ts_blk_start > valp->ts_ext4sync)
        data.ext4sync = valp->ts_blk_start - valp->ts_ext4sync;
    else if (valp->ts_ext4sync != 0) 
        data.ext4sync = ts_end - valp->ts_ext4sync;

    syscall_events.perf_submit(ctx, &data, sizeof(data));
    syscall_map.delete(&pid);

    // update histograms
    if (data.vfs)
        hist_vfs.increment(bpf_log2l(data.vfs / 1000));
    if (data.ext4)
        hist_ext4.increment(bpf_log2l(data.ext4 / 1000));
    if (data.writepg)
        hist_writepg.increment(bpf_log2l(data.writepg / 1000));
    if (data.ext4sync)
        hist_ext4sync.increment(bpf_log2l(data.ext4sync / 1000));
    if (data.blk)
        hist_blk.increment(bpf_log2l(data.blk / 1000));
    if (data.split)
        hist_split.increment(bpf_log2l(data.split / 1000));
    if (data.merge)
        hist_merge.increment(bpf_log2l(data.merge / 1000));
    
    return 0;
}

File: tools/ioflow/ioflow-syncwrite.py
#!/usr/bin/python
# 
# ioflow-syncwrite.py        End-to-end IO Flow Tracer for synchronous write() syscall. 
#
#
# Copyright (c) Google LLC
# Licensed under the Apache License, Version 2.0 (the "License")
#
# 27-Aug-2020   Haoning Chen   Created this. 

from __future__ import print_function
from bcc import BPF 
import argparse

# parse arguments
examples = """examples:
    ./ioflow-syncwrite.py              # Trace sync write io flow. Default time threshold: 1ms for syscalls and 0.2ms for requests.
    ./ioflow-syncwrite.py -s 5         # Set syscall threshold to 5ms. Print syscall data if its latency exceeds 5 ms.
    ./ioflow-syncwrite.py -s 5 -r 0.5  # Set syscall threshold to 5ms and request threshold to 0.5 ms.
    Any syscall data and request data that takes time longer than the corresponding threshold is emitted. 
"""
parser = argparse.ArgumentParser(
    description="End-to-end IO Flow Tracer for sync write IOs",
    formatter_class=argparse.RawDescriptionHelpFormatter,
    epilog=examples)
parser.add_argument("-s", "--sys_thres", type=float, default=1.,
    help="Set syscall threshold in ms. Emit any syscall data that takes time longer than this threshold. Default to be 1.0")
parser.add_argument("-r", "--rq_thres", type=float, default=0.2,
    help="Set request threshold in ms. Emit any request data that takes time longer than this threshold. Default to be 0.2")
args = parser.parse_args()


# load Tracer program 
with open("ioflow-syncwrite.c") as src_f:
    bpf_text = src_f.read()
with open("ioflow-common.h") as comm_f:
    comm_text = comm_f.read()

sys_thres_flag = "-DSYSCALL_THRESHOLD=%d" % int(args.sys_thres * 1000000)
rq_thres_flag = "-DREQUEST_THRESHOLD=%d" % int(args.rq_thres * 1000000)
# load BPF program
bpf = BPF(text=bpf_text.replace("[IMPORT_COMM]", comm_text, 1), cflags=[sys_thres_flag, rq_thres_flag])

# file system layer:
bpf.attach_kprobe(event="vfs_write", fn_name="vfs_write_entry")
bpf.attach_kprobe(event="ext4_file_write_iter", fn_name="ext4_entry")
bpf.attach_kprobe(event="generic_perform_write", fn_name="write_page_entry")
bpf.attach_kprobe(event="ext4_sync_file", fn_name="ext4_sync_entry")
# block layer:
bpf.attach_kprobe(event="submit_bio", fn_name="block_entry")
bpf.attach_kretprobe(event="submit_bio", fn_name="block_return")
bpf.attach_kprobe(event="bio_split", fn_name="split_entry")
bpf.attach_kretprobe(event="bio_split", fn_name="split_return")
bpf.attach_kprobe(event="bio_attempt_front_merge", fn_name="merge_entry")
bpf.attach_kretprobe(event="bio_attempt_front_merge", fn_name="merge_return")
bpf.attach_kprobe(event="bio_attempt_back_merge", fn_name="merge_entry")
bpf.attach_kretprobe(event="bio_attempt_back_merge", fn_name="merge_return")
# async request handling:
bpf.attach_kprobe(event="blk_account_io_start", fn_name="rq_create")
if BPF.get_kprobe_functions(b'blk_start_request'):
	bpf.attach_kprobe(event="blk_start_request", fn_name="rq_issue")
bpf.attach_kprobe(event="blk_mq_start_request", fn_name="rq_issue")
bpf.attach_kprobe(event="blk_account_io_done", fn_name="rq_done")
# end of IO:
bpf.attach_kretprobe(event="vfs_write", fn_name="vfs_write_return")


# header
print("Tracing sync write I/Os. Time threshold: %.1f ms for syscalls and %.1f ms for requests. Emit data that takes time longer than its threshold." % (args.sys_thres, args.rq_thres))
print("2 types of emit output with the following formats (unit: us):\n")

print("[REQUEST] %6s %6s %9s %9s %9s %11s %12s %8s %6s\n" % 
    ("PID", "IO_NUM", "TOTAL_LAT", "CREATE_TS", "QUEUE_LAT", "SERV_LAT", "SECTOR", "LEN", "DISK"))

print("[SYSCALL] %6s %6s %9s %9s %9s %11s %12s %9s %8s %8s %5s %11s %9s %9s %5s %11s %9s %9s %5s %10s %8s %8s %6s\n" % 
    ("PID", "IO_NUM", "TOTAL_LAT", "VFS_LAT", "EXT4_LAT", "WRITEPG_LAT", "EXT4SYNC_LAT", "BLK_START", 
    "BLK_LAT", "BLK_END", "COUNT", "SPLIT_START", "SPLIT_LAT", "SPLIT_END", "COUNT", "MERGE_START", 
    "MERGE_LAT", "MERGE_END", "COUNT", "COMMAND", "OFFSET", "SIZE", "FILE"))

print("Hit Ctrl-C to end and display histograms.\n")


# process syscall events
def print_syscall_event(cpu, data, size):
    event = bpf["syscall_events"].event(data)
    total = float(event.total) / 1000
    if total >= (args.sys_thres * 1000):
        ts_blk_start = float(event.ts_blk_start - event.ts_vfs) / 1000 if event.ts_blk_start > 0 else 0.
        ts_blk_end = float(event.ts_blk_end - event.ts_vfs) / 1000 if event.ts_blk_end > 0 else 0.
        ts_split_start = float(event.ts_split_start - event.ts_vfs) / 1000 if event.ts_split_start > 0 else 0.
        ts_split_end = float(event.ts_split_end - event.ts_vfs) / 1000 if event.ts_split_end > 0 else 0.
        ts_merge_start = float(event.ts_merge_start - event.ts_vfs) / 1000 if event.ts_merge_start > 0 else 0.
        ts_merge_end = float(event.ts_merge_end - event.ts_vfs) / 1000 if event.ts_merge_end > 0 else 0.

        print("[SYSCALL] %6s %6s %9.3f %9.3f %9.3f %11.3f %12.3f %9.3f %8.3f %8.3f %5s %11.3f %9.3f %9.3f %5s %11.3f %9.3f %9.3f %5s %10s %8s %8s %6s" 
            % (event.pid, event.seq_num, total, float(event.vfs)/1000, float(event.ext4)/1000, float(event.writepg)/1000, 
            float(event.ext4sync)/1000, ts_blk_start, float(event.blk)/1000, ts_blk_end, event.cnt_blk, 
            ts_split_start, float(event.split)/1000, ts_split_end, event.cnt_split, ts_merge_start, float(event.merge)/1000, 
            ts_merge_end, event.cnt_merge, event.cmd_name, event.offset, event.size, event.file_name))


# process request events
def print_rq_event(cpu, data, size):
    event = bpf["rq_events"].event(data)
    total = float(event.queue + event.service) / 1000
    if total >= (args.rq_thres * 1000):
        print("[REQUEST] %6s %6s %9.3f %9.3f %9.3f %11.3f %12s %8s %6s" % 
            (event.pid, event.seq_num, total, float(event.ts_create)/1000, float(event.queue)/1000, 
            float(event.service)/1000, event.sector, event.len, event.disk_name))


# loop with callback to print_event
bpf["syscall_events"].open_perf_buffer(print_syscall_event, page_cnt=64)
bpf["rq_events"].open_perf_buffer(print_rq_event, page_cnt=128)
while True:
    try:
        bpf.perf_buffer_poll()
    except KeyboardInterrupt:
        break

print()
bpf["hist_vfs"].print_log2_hist("VFS (us)")
print()
bpf["hist_ext4"].print_log2_hist("EXT4 (us)")
print()
bpf["hist_writepg"].print_log2_hist("Write Page (us)")
print()
bpf["hist_ext4sync"].print_log2_hist("EXT4 Sync (us)")
print()
bpf["hist_blk"].print_log2_hist("Block Entry (us)")
print()
bpf["hist_split"].print_log2_hist("Bio Split (us)")
print()
bpf["hist_merge"].print_log2_hist("Bio Merge (us)")
print()
bpf["hist_rq_queue"].print_log2_hist("Request Queue (us)")
print()
bpf["hist_rq_service"].print_log2_hist("Request Service (us)")
exit()
